Document how Keycloak is upgraded when Operator is upgraded via OLM (#38297)
keycloak/keycloak: docs/guides/operator/installation.adoc

* Be sure that you have cluster-admin permission or an equivalent level of permissions granted by an administrator.

> [WARNING]

Reviewer: This structure looks a bit weird to me:
<img width="1085" alt="Screenshot 2025-03-27 at 17 07 24" src="https://github.com/user-attachments/assets/2b445512-010a-42ae-8164-d5031b001d73" />
Maybe we could swap the warning box with the "Configuring Manual Approval for OLM Upgrades" heading so it's under that?

====
*Important: Automatic OLM Upgrades*

By default, OLM automatically updates the {project_name} Operator when a new version is released. This can cause several significant issues:

* When using the default {project_name} image, the Operator by default uses a matching image of the corresponding {project_name} version, resulting in *unintended {project_name} upgrades* when the Operator is upgraded.

> * Even when using custom images, major Operator upgrades can introduce compatibility issues with your existing Keycloak CR configuration.

Reviewer 2: Ideally it's only on major releases. There could be a word of caution on minor releases as well because of operator logic changes.

*Potential risks:*

* Unexpected downtime during automatic upgrades

> * Database schema migrations that cannot be easily rolled back

Reviewer 2: In particular based upon recent issues, you cannot even reliably simply roll back micro version changes - that is once 26.x.n is applied, if you want to go back to 26.x.(n -1) then you need to restore from a database snapshot, rather than just going back 1 micro in operator / keycloak versions.

Reviewer: We could possibly reword it as:
```suggestion
* No option to downgrade to the previous {project_name} version due to database schema migrations
```
@Pepo48 @shawkins WDYT?

> * Version incompatibilities if custom resources use features from newer CRDs

Reviewer: Not sure what you mean by this?

* Breaking changes in the Operator behavior across major versions

*Recommendation:*

*We strongly recommend using only manual approval mode for the Keycloak Operator.* This ensures you can:

1. Schedule maintenance windows for upgrades
2. Test upgrades in a non-production environment first

> 3. Prepare rollback plans in case of issues

Reviewer 2: May need to elaborate here - or at least link to the general recommendations on changing versions, which mention backing up the database.

Reviewer: We do not have specific backup instructions in the [upgrading guide](https://www.keycloak.org/docs/latest/upgrading/index.html#_upgrading). Still, we might improve the wording, something like:
```suggestion
3. Back up the database to allow downgrading to the previous {project_version} in case of issues
```

Author: @shawkins do you have specific section/guide to link here in mind?

> 4. Review release notes and understand changes before approving upgrades

Reviewer: IMHO this should be the very first step.

> See the instructions below for configuring manual approval with OLM.

Reviewer: IMHO no need for this sentence if it's at the beginning of the "Configuring Manual Approval for OLM Upgrades" chapter.

====

> ==== Configuring Manual Approval for OLM Upgrades

Reviewer: I'd move this after the "Using the OpenShift web console" chapter,  at the end of "Installing by using the Operator Lifecycle Manager". It IMHO makes more sense there structure wise.

You may select to either have the Operator watch the namespace where it is installed, or to watch a single namespace of your choosing.

==== Configuring Manual Approval for OLM Upgrades

[WARNING]
====
*Important: Automatic OLM Upgrades*

By default, OLM automatically updates the {project_name} Operator when a new version is released. This can cause several significant issues:

* When using the default {project_name} image, the Operator uses a matching image of the corresponding {project_name} version, resulting in *unintended {project_name} upgrades* when the Operator is upgraded
* *Even when using custom images*, major Operator upgrades can introduce significant compatibility issues with your existing Keycloak CR configuration, potentially requiring manual intervention
* New fields in Keycloak CR or behavioral changes could impact existing deployments
* No option to downgrade to the previous {project_name} version due to changes related to database migration

*Recommendation:*

*We strongly recommend using manual approval mode for the Keycloak Operator.* This ensures you can:

> 1. Review release notes and understand changes before approving upgrades

Reviewer: ```suggestion
1. Review release notes and follow migration changes before approving upgrades
```

2. Schedule maintenance windows for upgrades
3. Test upgrades in a non-production environment first

> 4. Back up the database to allow downgrading to the previous {project_version} in case of issues

Reviewer: Noticed a misspelled variable.
```suggestion
4. Back up the database to allow downgrading to the previous {project_name} in case of issues
```

You may select to either have the Operator watch the namespace where it is installed, or to watch a single namespace of your choosing.

==== Configuring Manual Approval for OLM Upgrades

[WARNING]
====
*Important: Automatic OLM Upgrades*

By default, OLM automatically updates the {project_name} Operator when a new version is released. This can cause several significant issues:

*Major version upgrades* present the highest risk:

* When using the default {project_name} image, the Operator uses a matching image of the corresponding {project_name} version, resulting in *unintended {project_name} upgrades* when the Operator is upgraded
* *Even when using custom images*, major Operator upgrades can introduce significant compatibility issues with your existing Keycloak CR configuration, potentially requiring manual intervention

> * No option to downgrade to the previous {project_name} version due to database schema migrations

Reviewer: I would not split it between major and minor upgrades. All points are common for both.

Author: > Ideally it's only on major releases. There could be a word of caution on minor releases as well because of operator logic changes.

Just to somehow address Steven's concern above, which I think was valid, I can put it together as you suggest and add the following caution remark:
`These issues are particularly pronounced during major version upgrades, but can also occur in minor releases.`
wdyt?

Reviewer 2: > No option to downgrade to the previous {project_name} version due to database schema migrations

Not necessarily because of schema migrations, but other changes in the migration scripts, in rare circumstances this affects micro releases as well.

Reviewer: Database migrations certainly happen regularly not only in major releases but in minor too. My suggestion of the overall wording might be something like this:
```
By default, OLM automatically updates the {project_name} Operator when a new version is released. This can cause several significant issues:

* When using the default {project_name} image, the Operator uses a matching image of the corresponding {project_name} version, resulting in *unintended {project_name} upgrades* when the Operator is upgraded
* *Even when using custom images*, major Operator upgrades can introduce significant compatibility issues with your existing Keycloak CR configuration, potentially requiring manual intervention
* New fields in Keycloak CR or behavioral changes could impact existing deployments
* No option to downgrade to the previous {project_name} version due to changes related to database migration
```

@Pepo48 @shawkins WDYT?

Author: Sounds fine for me, meanwhile I applied your suggestion.

> * Breaking changes in the Operator's handling of Custom Resources

Reviewer: I don't think we need this since we already have:
```
*Even when using custom images*, major Operator upgrades can introduce significant compatibility issues with your existing Keycloak CR configuration, potentially requiring manual intervention
```

Author: Agree. I will remove it.
